src/compiler/args.rs

```diff
@@ -1,21 +1,51 @@
 use std::cmp::Ordering;
+use std::error::Error;
 use std::ffi::OsString;
+use std::fmt::{self, Debug, Display};
+use std::marker::PhantomData;
 use std::path::PathBuf;
+use std::result::Result as StdResult;
+
+type ArgResult<T> = StdResult<T, ArgError>;
+
+#[derive(Debug, PartialEq)]
+pub enum ArgError {
+    UnexpectedEndOfArgs,
+}
+
+impl ArgError {
+    pub fn static_description(&self) -> &'static str {
+        match self {
+            ArgError::UnexpectedEndOfArgs => "Unexpected end of args",
+        }
+    }
+}
+
+impl Display for ArgError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = self.static_description();
+        write!(f, "{}", s)
+    }
+}
+
+impl Error for ArgError {
+    fn cause(&self) -> Option<&Error> { None }
+}
 
 pub type Delimiter = Option<u8>;
 
 /// Representation of a parsed argument
 #[derive(PartialEq, Clone, Debug)]
-pub enum Argument {
+pub enum Argument<T> {
     /// Unknown non-flag argument ; e.g. "foo"
     Raw(OsString),
     /// Unknown flag argument ; e.g. "-foo"
     UnknownFlag(OsString),
     /// Known flag argument ; e.g. "-bar"
-    Flag(&'static str),
+    Flag(&'static str, T),
```

Comment:
It took me a lot more code reading to figure out why `Flag` needed to store the additional value here. I think the comment above `Argument<T>` could stand an additional sentence or two explaining what the type parameter is used for.